src/utils/AssayGeneratorMetabo.cpp

```diff
@@ -0,0 +1,523 @@
+        adduct = "["+adduct_prefix+"]"+adduct_suffix;
+      }
+
+      // check if known unknown should be used
+      if (description == "UNKNOWN" && sumformula == "UNKNOWN" && adduct == "UNKNOWN" && !use_known_unknowns) { continue; }
+
+      double spectrum_rt = 0.0;
+      double highest_precursor_mz = 0.0;
+      float highest_precursor_int = 0.0;
+      MSSpectrum highest_precursor_int_spectrum;
+      MSSpectrum transition_spectrum;
+
+      // find precursor/spectrum with highest intensity precursor
+      std::vector<size_t> index = it->second;
+      for (std::vector<size_t>::iterator index_it = index.begin(); index_it != index.end(); ++index_it)
+      {
+        const MSSpectrum &spectrum = spectra[*index_it];
+        const vector<Precursor> &precursor = spectrum.getPrecursors();
+
+        // get m/z and intensity of precursor
+        double precursor_mz = precursor[0].getMZ();
+        float precursor_int = precursor[0].getIntensity();
+
+        // spectrum with highest intensity precursor
+        if (precursor_int > highest_precursor_int)
+        {
+          highest_precursor_int = precursor_int;
+          highest_precursor_mz = precursor_mz;
+          highest_precursor_int_spectrum = spectra[*index_it];
+          spectrum_rt = spectra[*index_it].getRT();
+        }
+        transition_spectrum = highest_precursor_int_spectrum;
+      }
+
+      // if only one MS2 is available and the consensus method is used - jump right to the transition list calculation
+      // fallback: highest intensity precursor
+      if (method_consensus_spectrum && index.size() >= 2)
+      {
+        // transform to binned spectra
+        std::vector<BinnedSpectrum> binned;
+        std::vector<MSSpectrum> similar_spectra;
+        MSExperiment exp;
+        const BinnedSpectrum binned_highest_int(highest_precursor_int_spectrum, BinnedSpectrum::DEFAULT_BIN_WIDTH_HIRES, false, 1);
+
+        // calculation of contrast angle (cosine simiarity)
+        for (std::vector<size_t>::iterator index_it = index.begin(); index_it != index.end(); ++index_it)
+        {
+          const MSSpectrum &spectrum = spectra[*index_it];
+          const BinnedSpectrum binned_spectrum(spectrum, BinnedSpectrum::DEFAULT_BIN_WIDTH_HIRES, false, 1);
+
+          BinnedSpectralContrastAngle bspa;
+          double cosine_sim = bspa(binned_highest_int, binned_spectrum);
+
+          if (cosine_sim > cosine_sim_threshold)
+          {
+            similar_spectra.push_back(spectrum);
+            exp.addSpectrum(spectrum);
+          }
+        }
+        // calculate consensus spectrum
+        exp.sortSpectra();
+        SpectraMerger merger;
+        Param p;
+        p.setValue("precursor_method:mz_tolerance", precursor_mz_tol*2);
+        p.setValue("precursor_method:rt_tolerance", precursor_rt_tol*2);
+        merger.setParameters(p);
+
+        // all MS spectra should have the same precursor
+        merger.mergeSpectraPrecursors(exp);
```

Comment:
this is a good start but we will have to look into whether this is really the optimal way of averaging / merging spectra